### PR TITLE
feat(http): maxRequestsPerSecond

### LINKED ISSUE #6359: Support maxRequestsPerSecond
**What would you like Renovate to be able to do?**

Allow configuration of a maximum requests per X, e.g. 5 requests per second.

**Describe the solution you'd like**

hostRules

### PR DESCRIPTION
**Pull request by zharinov**
## Context

- Closes #6359

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — CHANGES_REQUESTED** [review]
Can we simplify by doing RPS? (and allowing <1, e.g. 0.5 RPS)

File: lib/util/http/index.ts [hunk 1]
```diff
@@ -162,24 +163,32 @@ export class Http<Opts extends HttpOptions = HttpOptions> {
...
+      resPromise = queuedTask();
       if (options.method === 'get' || options.method === 'head') {
         memCache.set(cacheKey, resPromise); // always set if it's a get or a head
       }
     }
 
     try {
       const res = await resPromise;
-      res.authorization = !!options?.headers?.authorization;
+      (res as any).authorization = !!options?.headers?.authorization;
```
Comment (viceice) [inline lib/util/http/index.ts:191]: we should cast res instead

File: lib/util/http/index.ts [hunk 2]
```diff
@@ -62,7 +63,7 @@ function applyDefaultHeaders(options: Options): void {
 // `Buffer` in the latter case.
 // We don't declare overload signatures because it's immediately wrapped by
 // `request`.
-async function gotRoutine<T>(
+async function gotTask<T>(
```
Comment (viceice) [inline lib/util/http/index.ts:66]: can we do the renaming in a separate PR?

File: lib/util/http/throttle.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import { parseUrl } from '../url';
+import { getThrottleInterval, getThrottleLimit } from './host-rules';
+
+const hostThrottles = new Map<string, Throttle | null>();
+
+class Throttle {
+  constructor(
+    limit: number,
+    interval: number,
+    private throttle = pThrottle({
```
Comment (viceice) [inline lib/util/http/throttle.ts:13]: does it need to be a constructor arg?

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/util/http/index.ts (hunk 1 above)
Comment (zharinov) [reply lib/util/http/index.ts:191]: To which type?

Comment (viceice) [reply lib/util/http/index.ts:191]: should probably `HttpResponse<T>` ? it's the function return type.

Comment (zharinov) [reply lib/util/http/index.ts:194]: Hah, good catch

File: lib/util/http/throttle.ts (hunk 3 above)
Comment (zharinov) [reply lib/util/http/throttle.ts:13]: This would lead to way more ugly types. I think it's better to rely on inferred types here.

Comment (viceice) [reply lib/util/http/throttle.ts:13]: i don't see any generic here, so i don't understand what the problem with a static type. is there no definition for the return value?

File: lib/config/options/index.ts [hunk 4]
```diff
@@ -2107,6 +2107,16 @@ const options: RenovateOptions[] = [
     cli: false,
     env: false,
   },
+  {
+    name: 'maxRequestsPerSecond',
+    description: 'Limit requests rate per host.',
+    type: 'integer',
+    stage: 'repository',
+    parent: 'hostRules',
+    default: null,
```
Comment (rarkins) [inline lib/config/options/index.ts:2116]: Will `null` be the same meaning as `0`?

File: lib/util/http/host-rules.ts [hunk 5]
```diff
@@ -120,10 +121,23 @@ export function applyHostRules(url: string, inOptions: GotOptions): GotOptions {
...
+export function getThrottleOptions(
+  url: string
+): { limit: number; interval: number } | null {
+  const hostRule = hostRules.find({ url });
+  const throttleLimit = hostRule.maxRequestsPerSecond;
+  if (!is.number(throttleLimit)) {
+    return null;
+  }
+  const limit = Math.floor(throttleLimit * 1000);
+  const interval = 1000 * 1000;
```
Comment (rarkins) [inline lib/util/http/host-rules.ts:141]: I think the term "interval" usually means the gap between things, whereas here you seem to be using it as the time window, so I think terminology should be changed.

If we were to implement this simply, meaning "enforce the gap between each request" then it would be simpler, a well as more likely to satisfy other parties. e.g. if a party says maximum 10 RPS but we send 10 in the first millisecond, then wait 999ms, then send 10, etc - they likely won't be happy.

So what if you *actually* return the interval? e.g. if 1 RPS the interval is 1000ms, if 10RPS then 100ms, if it's 0.5RPS then the interval is 2000ms, etc.

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/http/host-rules.ts (hunk 5 above)
Comment (zharinov) [reply lib/util/http/host-rules.ts]: I think I finally implemented your idea

File: lib/config/options/index.ts (hunk 4 above)
Comment (zharinov) [reply lib/config/options/index.ts:2116]: Yes

**Review by rarkins — DISMISSED** [review]

File: lib/config/options/index.ts [hunk 6]
```diff
@@ -2107,6 +2107,16 @@ const options: RenovateOptions[] = [
     cli: false,
     env: false,
   },
+  {
+    name: 'maxRequestsPerSecond',
+    description: 'Limit requests rate per host.',
+    type: 'integer',
```
Comment (viceice) [inline lib/config/options/index.ts:2113]: 🤔 does the validator accept float for this type?

File: lib/util/http/host-rules.ts [hunk 7]
```diff
@@ -120,10 +121,14 @@ export function applyHostRules(url: string, inOptions: GotOptions): GotOptions {
...
 }
 
-export function getRequestLimit(url: string): number | null {
-  const hostRule = hostRules.find({
-    url,
-  });
-  const limit = hostRule.concurrentRequestLimit;
+export function getConcurrentRequestsLimit(url: string): number | null {
+  const { concurrentRequestLimit: limit } = hostRules.find({ url });
   return typeof limit === 'number' && limit > 0 ? limit : null;
```
Comment (viceice) [inline lib/util/http/host-rules.ts:126]:
```suggestion
   return typeof limit === 'number' && limit > 0 ? limit : null;
```
we should use is.number here too,like below to unify code

File: lib/util/http/throttle.spec.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
+import * as hostRules from '../host-rules';
+import { clear, getThrottle } from './throttle';
+
+describe('util/http/throttle', () => {
+  beforeEach(() => {
+    jest
+      .spyOn(hostRules, 'find')
+      .mockReturnValue({ maxRequestsPerSecond: 143 });
+    clear();
+  });
```
Comment (viceice) [inline lib/util/http/throttle.spec.ts:10]: why not simply set a hostrule ? 😉

File: lib/util/http/throttle.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
+import pThrottle from 'p-throttle';
+import { logger } from '../../logger';
+import { parseUrl } from '../url';
+import { getThrottleIntervalMs } from './host-rules';
+
+const hostThrottles = new Map<string, Throttle | null>();
+
+class Throttle {
+  private throttle: <Argument extends readonly unknown[], ReturnValue>(
```
Comment (viceice) [inline lib/util/http/throttle.ts:9]: isn't this return type of pThrottle ?

File: lib/config/options/index.ts (hunk 6 above)
Comment (rarkins) [reply lib/config/options/index.ts:2113]: good question.. we need that

File: lib/util/http/throttle.ts (hunk 9 above)
Comment (zharinov) [reply lib/util/http/throttle.ts:9]: Yes, it is. Why do you ask?

Comment (zharinov) [reply lib/util/http/throttle.ts]: Never mind

File: lib/config/options/index.ts (hunk 6 above)
Comment (zharinov) [reply lib/config/options/index.ts:2113]: It supports fractional values, despite of being named as `integer`

Comment (viceice) [reply lib/config/options/index.ts:2106]: ok, then we should rename this to `number` later

**Review by viceice — APPROVED** [review]

File: lib/config/options/index.ts (hunk 6 above)
Comment (viceice) [reply lib/config/options/index.ts:2106]: this seems to be the only usage of string `integer` for `type`
https://github.com/renovatebot/renovate/blob/2349e5597626dc230394f00ffddf186c19bde7d9/lib/config/validation.ts#L258-L266

**Comment by Tranquility** [comment]
Does this unblock https://github.com/renovatebot/renovate/issues/3670?
